src/include/migraphx/rewrite_reshapes.hpp

```diff
@@ -0,0 +1,206 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+#ifndef MIGRAPHX_GUARD_MIGRAPHX_REWRITE_RESHAPES_HPP
+#define MIGRAPHX_GUARD_MIGRAPHX_REWRITE_RESHAPES_HPP
+
+#include <migraphx/config.hpp>
+#include <migraphx/matcher.hpp>
+#include <migraphx/make_op.hpp>
+#include <migraphx/pass_manager.hpp>
+#include <migraphx/common_dims.hpp>
+#include <migraphx/simplify_reshapes.hpp>
+#include <migraphx/eliminate_common_subexpression.hpp>
+#include <migraphx/dead_code_elimination.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+struct rewrite_reshapes_base
+{
+    template <class AxesMap>
+    static instruction_ref insert(module_pass_manager& mpm,
+                                  instruction_ref ins,
+                                  const std::vector<instruction_ref>& inputs,
+                                  const AxesMap&)
+    {
+        return mpm.get_module().insert_instruction(
+            ins, ins->get_operator(), inputs, ins->module_inputs());
+    }
+
+    template <class AxesMap>
+    static bool supports(instruction_ref, std::vector<std::size_t>&, const AxesMap&)
+    {
+        return true;
+    }
+
+    static std::vector<std::size_t> base_dims(instruction_ref ins)
+    {
+        return ins->get_shape().lens();
+    }
+};
+
+template <class T>
+struct rewrite_reshapes
+{
+    std::string name() const { return "rewrite_reshapes"; }
+    struct find_op_reshape_op
+    {
+        std::string op1;
+        std::string op2;
+
+        auto matcher() const
+        {
+            auto reshape =
+                match::name("reshape", "squeeze", "unsqueeze", "flatten")(match::used_once());
+            auto skip_contiguous = [](auto... ms) {
+                return match::arg(0)(match::skip(
+                    match::name("contiguous", "multibroadcast")(match::used_once()))(ms...));
+            };
+            auto pointwise         = match::name(op1)(match::used_once());
+            auto reshape_pointwise = reshape(skip_contiguous(pointwise.bind("x"))).bind("reshape");
+            return match::name(op2)(match::any_of[match::inputs()](reshape_pointwise));
+        }
+
+        template <class F>
+        static instruction_ref find_input_if(instruction_ref start, instruction_ref last, F f)
+        {
+            while(start != last)
+            {
+                if(f(start))
+                    return start;
+                if(start->inputs().size() != 1)
```

Comment:
Add additional test case for this early return condition.  Part of me wants to say make this a break instead as you're returning the same thing as the  outside of the loop.

Reply:
Looks like there is coverage for that. Also early return seems to make it clearer what will happen when this condition is met rather having to follow where the break will end up.

Reply:
fair point

Reply:
There isnt coverage for this line. We actually wont hit this with the way its currently called, but if this was called in a different context we wouldn't want it to keep traversing here which is why the condition is here.

Reply:
I understand all good.